File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.java
import java.io.*;
import java.util.*;


public class Main {
    static List<Integer> list[];
    static boolean visited[];
    static int arr[];

    static void BFS(int start){
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        int count = 0;
        visited[start] = true;
        while (!queue.isEmpty()){
            int now = queue.poll();

            for (int next: list[now]){
                if (visited[next]) continue;
                visited[next] = true;
                arr[next]++;
                queue.add(next);

            }
        }
    }
    public static void main(String[] args) throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        list = new ArrayList[n + 1];
        arr = new int[n + 1];
        for (int i = 1; i<=n;i++) list[i] = new ArrayList<>();

        for (int i = 0;i<m;i++){
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            list[a].add(b);
        }
        int max = 0;
        for (int i = 1;i<=n;i++){
            visited = new boolean[n + 1];
            BFS(i);
        }
        for (int i = 1;i<=n;i++) max = Math.max(max, arr[i]);

        StringBuilder sb = new StringBuilder();
        for (int i = 1;i<=n;i++){
            if (max == arr[i]) sb.append(i + " ");
        }
        System.out.println(sb);
    }
}

File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.cc
#include <iostream>
#include <vector>
#include <queue>
#include <cstring>
using namespace std;

int n, m;
int ans[10001];
bool visited[10001];

int BFS(const vector<vector<int>>& arr, int idx) {
    queue<int> q;
    memset(visited, false, sizeof(visited));

    visited[idx] = true;
    q.push(idx);

    int cnt = 1;

    while (!q.empty()) {
        int now = q.front();
        q.pop();

        for (int next : arr[now]) {
            if (visited[next]) continue;

            visited[next] = true;
            q.push(next);
            cnt++;
        }
    }

    return cnt;
}

int main(void) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n >> m;
    vector<vector<int>> arr(n + 1);

    for (int i = 0; i < m; i++) {
        int a, b; cin >> a >> b;
        arr[b].push_back(a); // 역방향
    }

    int maxs = 0;

    for (int i = 1; i <= n; i++) {
        ans[i] = BFS(arr, i);
        maxs = max(maxs, ans[i]);
    }

    for (int i = 1; i <= n; i++) {
        if (ans[i] == maxs)
            cout << i << " ";
    }
}